engines/sci/sound/audio32.cpp

```diff
@@ -0,0 +1,920 @@
+		}
+
+		if (_attenuatedMixing) {
+			leftVolume >>= attenuationAmount;
+			rightVolume >>= attenuationAmount;
+		}
+
+		if (channelIndex == _monitoredChannelIndex) {
+			const size_t bufferSize = numSamples * sizeof(Audio::st_sample_t);
+			if (_monitoredBufferSize < bufferSize) {
+				_monitoredBuffer = (Audio::st_sample_t *)realloc(_monitoredBuffer, bufferSize);
+				_monitoredBufferSize = bufferSize;
+			}
+
+			memset(_monitoredBuffer, 0, _monitoredBufferSize);
+
+			_numMonitoredSamples = writeAudioInternal(channel.stream, channel.converter, _monitoredBuffer, numSamples, leftVolume, rightVolume, channel.loop);
+
+			Audio::st_sample_t *sourceBuffer = _monitoredBuffer;
+			Audio::st_sample_t *targetBuffer = buffer;
+			const Audio::st_sample_t *const end = _monitoredBuffer + _numMonitoredSamples;
+			while (sourceBuffer != end) {
+				Audio::clampedAdd(*targetBuffer++, *sourceBuffer++);
+			}
+
+			if (_numMonitoredSamples > maxSamplesWritten) {
+				maxSamplesWritten = _numMonitoredSamples;
+			}
+		} else if (!channel.stream->endOfStream() || channel.loop) {
+			int channelSamplesWritten = writeAudioInternal(channel.stream, channel.converter, buffer, numSamples, leftVolume, rightVolume, channel.loop);
+			if (channelSamplesWritten > maxSamplesWritten) {
+				maxSamplesWritten = channelSamplesWritten;
+			}
+		}
+	}
+
+	return maxSamplesWritten;
+}
+
+#pragma mark -
+#pragma mark Channel management
+
+int16 Audio32::findChannelByArgs(int argc, const reg_t *argv, const int startIndex, const reg_t soundNode) const {
+	// NOTE: argc/argv are already reduced by one in our engine because
+	// this call is always made from a subop, so no reduction for the
+	// subop is made in this function. SSCI takes extra steps to skip
+	// the subop argument.
+
+	argc -= startIndex;
+	if (argc <= 0) {
+		return kAllChannels;
+	}
+
+	Common::StackLock lock(_mutex);
+
+	if (_numActiveChannels == 0) {
+		return kNoExistingChannel;
+	}
+
+	// This is a plain Audio resource
+	if (argc < 5) {
+		const ResourceId searchId(kResourceTypeAudio, argv[startIndex].toUint16());
+
+		for (int16 i = 0; i < _numActiveChannels; ++i) {
+			const AudioChannel &candidate = getChannel(i);
+
+			if (
+				candidate.id == searchId &&
+				(soundNode.isNull() || candidate.soundNode == soundNode)
+			) {
+				return i;
+			}
+		}
+	// This is an Audio36 resource
+	} else {
+		const ResourceId searchId(kResourceTypeAudio36, argv[startIndex].toUint16(), argv[startIndex + 1].toUint16(), argv[startIndex + 2].toUint16(), argv[startIndex + 3].toUint16(), argv[startIndex + 4].toUint16());
+
+		for (int16 i = 0; i < _numActiveChannels; ++i) {
+			const AudioChannel &candidate = getChannel(i);
+
+			if (candidate.robot == false && candidate.id == searchId) {
+				return i;
+			}
+		}
+	}
+
+	return kNoExistingChannel;
+}
+
+int16 Audio32::findChannelById(const ResourceId resourceId, const reg_t soundNode) const {
```

Comment:
Thanks for noticing this; while `findChannelById` is used only by sound effects calls from `kDoSound` right now, it should be usable regardless of the audio resource type, especially since later engine versions change kernel calling conventions which means `findChannelById` is probably going to be used more frequently later.